File: src/main/java/com/prmncr/normativecontrol/serializers/ByteArraySerializer.java
package com.prmncr.normativecontrol.serializers;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;

import java.io.IOException;

public class ByteArraySerializer extends JsonSerializer<byte[]> {

    @Override
    public void serialize(byte[] bytes, JsonGenerator gen, SerializerProvider provider) throws IOException {
        gen.writeStartArray();
        for (byte b : bytes) {
            gen.writeNumber(unsignedToBytes(b));
        }
        gen.writeEndArray();
    }

    private static int unsignedToBytes(byte b) {
        return b & 0xFF;
    }
}

File: src/main/kotlin/com/prmncr/normativecontrol/serializers/ByteArraySerializer.kt
package com.prmncr.normativecontrol.serializers

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider

class ByteArraySerializer : JsonSerializer<ByteArray>() {
    override fun serialize(bytes: ByteArray, gen: JsonGenerator, provider: SerializerProvider) {
        val hexChars = CharArray(bytes.size * 2)
        for (j in bytes.indices) {
            val v = bytes[j].toInt() and 0xFF
            hexChars[j * 2] = HEX_ARRAY[v ushr 4]
            hexChars[j * 2 + 1] = HEX_ARRAY[v and 0x0F]
        }
        gen.writeString(String(hexChars))
    }

    companion object {
        private val HEX_ARRAY =
            charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
    }
}